Add fan mode and swing mode support to toshiba climate

casch-at opened the pull request:
[reverse_engineered_toshiba_ir_frames.txt](https://github.com/esphome/esphome/files/6299796/reverse_engineered_toshiba_ir_frames.txt)

# What does this implement/fix? 

This change adds support for fan mode and swing mode to the toshiba climate component. 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes https://github.com/esphome/issues/issues/1752

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1167
  
# Test Environment

- [x] ESP32
- [x] ESP8266
- [ ] Windows
- [ ] Mac OS
- [x] Linux

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
remote_receiver:
  id: rcvr
  pin:
    number: 32
    mode: INPUT_PULLUP
    inverted: True
  dump: all

remote_transmitter:
  id: tsmtr
  pin: 27
  carrier_duty_percent: 50%

climate:
 - platform: toshiba
   id: toshiba_ac
   name: 'Toshiba AC'
   transmitter_id: tsmtr
   receiver_id: rcvr

switch:
  - platform: template
    name: 'Toshiba AC Preset'
    id: toshiba_ac_preset
    icon: mdi:cached
    turn_on_action:
      - climate.control:
          id: toshiba_ac
          mode: COOL
          target_temperature: 24°C
          fan_mode: AUTO
          swing_mode: 'OFF'

```

# Explain your changes
This change logs the Toshiba frame (receive/transmit) for easier debugging (line 55). I'm not sure if it makes sense to add this to `esphome/components/remote_base/toshiba_protocol.cpp`? 
`transmit_state` will send two commands, if swing mode changes too, beside mode, fan mode or temperature. This is a little annoying in my opinion, because the unit beeps two time in a row. However, we could use the default command (0x01), and add two bytes of extra information, see section 4 Preset Button in the attached file. But I don't know if all AC units support that, further, at the moment I don't know what 0x21 represents in the data field, but they seem to be always unchanged regardless of the command sent.

Attached you can find text file with the reverse engineered IR frames. 

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

# Attachments
[reverse_engineered_toshiba_ir_frames.txt](https://github.com/esphome/esphome/files/6299798/reverse_engineered_toshiba_ir_frames.txt)

casch-at committed:
Add fan mode and swing mode support to toshiba climate

casch-at committed:
Fix all errors/warnings

JoppyFurr commented on esphome/components/toshiba/toshiba.cpp:
```diff
@@ -11,61 +14,77 @@ const uint16_t TOSHIBA_ZERO_SPACE = 540;
+const uint8_t TOSHIBA_FAN_SPEED_2 = 0x06;
+const uint8_t TOSHIBA_FAN_SPEED_3 = 0x08;
+const uint8_t TOSHIBA_FAN_SPEED_4 = 0x0a;
+const uint8_t TOSHIBA_FAN_SPEED_5 = 0x0c;
```
Any reason you're right shifting these values, only to left-shift them back again later?

JoppyFurr commented on esphome/components/toshiba/toshiba.cpp:
```diff
@@ -11,61 +14,77 @@ const uint16_t TOSHIBA_ZERO_SPACE = 540;
 
-static const char *TAG = "toshiba.climate";
+const uint8_t TOSHIBA_MOTION_FIX = 0x00;
+const uint8_t TOSHIBA_MOTION_SWING = 0x01;
```
Maybe our devices are a little different?

Investigating my remote, data[5] of the packet has the following two values when pressing swing/fix:
* 0x04 when pressing the "Swing" button (toggles the swinging motion on and off)
* 0x00 when pressing the "Fix" button (moves the airflow by a couple of degrees each press)

My device does not have a "SWING_OFF" value, instead the remote sends identical frames for on and off:

```
Swing:      0x f2 0d 01 fe  21 04 25
Swing_off:  0x f2 0d 01 fe  21 04 25
Fix:        0x f2 0d 01 fe  21 00 21
```

I'll attach a picture of my remote, to see if we have slightly different models.
It may be worth making the swing/fix codes, and any other possible differences, configurable through yaml.

![remote_front](https://user-images.githubusercontent.com/1698231/114630825-7ea8e400-9d0f-11eb-8b27-8d0409b0fd12.jpg)
![remote_back](https://user-images.githubusercontent.com/1698231/114630836-84062e80-9d0f-11eb-959c-cd10a76af3ec.jpg)

casch-at committed:
Fix formatting and cast return value of clamp

casch-at committed:
Keep fan mode left shifted by 4

casch-at committed:
Add horizontal and both swing mode

casch-at committed:
Add horizontal and both swing mode

cslux commented:
@JoppyFurr Sorry for pinging you. But I would need something from you! Could you please test if 0x02 works for turning off the swing mode. I'm asking because for my remote both 0x01 and 0x04 would work for vertical swing mode. If 0x02 works for you, than I don't need to distinguish between the different remote. The only thing which needs to be customizable than is whether the remote supports additionally BOTH and HORIZONTAL. Thank you!

casch-at committed:
Add support for different models

JoppyFurr commented:
> @JoppyFurr Sorry for pinging you. But I would need something from you! Could you please test if 0x02 works for turning off the swing mode. I'm asking because for my remote both 0x01 and 0x04 would work for vertical swing mode. If 0x02 works for you, than I don't need to distinguish between the different remote. The only thing which needs to be customizable than is whether the remote supports additionally BOTH and HORIZONTAL. Thank you!

Hi @cslux ,

I will give this a try later today and will get back to you,

Cheers,
Joppy

JoppyFurr commented:
Hi @cslux ,

I have just tested, and yes, my heat pump works with the 0x01 / 0x02 codes too:

* 0x01: Always turns on vertical-swing
* 0x02: Always turns off vertical-swing
* 0x04: Toggles vertical-swing

Cheers,
Joppy

casch-at committed:
Add generic model and WH_TA01LE

Generic model supports vertical/fix/off swing mode
WH_TA01LE supports additionally horizontal/both swing mode, further the
numbers sent are different too.

casch-at committed:
Use MODEL_1 and MODEL_2 as model name

OttoWinter committed:
Merge remote-tracking branch 'origin/dev' into expand_toshiba_climate

WeekendWarrior1 commented:
@jesserockz has implemented "set_supported_swing_modes" in climate_ir.h here: [/1998/commits/531f5f7b7066a913e638913118eb72938889ef1b](https://github.com/esphome/esphome/pull/1998/commits/531f5f7b7066a913e638913118eb72938889ef1b), which may be of interest for this merge request.

cslux commented:
Thank you, @WeekendWarrior1 I will take a look at it next week when I have vacation.
